In [1]:
import pandas as pd
import numpy as np

In [2]:
from sklearn import datasets
from sklearn import metrics
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

# 분류 : 붓꽃 데이터

In [3]:
iris = pd.read_csv('붓꽃 데이터.csv')
iris.head()

caseno  SepalLength  SepalWidth  PetalLength  PetalWidth Species
0       1          5.1         3.5          1.4         0.2  setosa
1       2          4.9         3.0          1.4         0.2  setosa
2       3          4.7         3.2          1.3         0.2  setosa
3       4          4.6         3.1          1.5         0.2  setosa
4       5          5.0         3.6          1.4         0.2  setosa

In [4]:
#Species를 0,1,2로 바꿔줌.
tmp_species = []
for species in iris['Species']:
    if species == 'setosa':
        tmp_species.append(0)
    if species == 'versicolor':
        tmp_species.append(1)
    if species == 'virginica':
        tmp_species.append(2)

tmp_species

[0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2]

In [5]:
#상수로 변환한 값을 원래 Species 열에 할당
iris['Species'] = tmp_species

In [6]:
#iris의 독립변수와 종속변수 설정

iris_data = iris[['SepalLength', 'SepalWidth', 'PetalLength', 'PetalWidth']]
iris_label = iris['Species']

In [7]:
#train_test_split을 활용하여 독립변수와 종속변수를 할당

X_train, X_test, y_train, y_test = train_test_split(iris_data, iris_label, test_size = 0.2, random_state = 42)

### DecisionTreeClassifier

In [8]:
#DecisionTreeClassifier를 활용하여 학습

dt_model = DecisionTreeClassifier(max_depth = 3, random_state = 42)
dt_model.fit(X_train, y_train)

DecisionTreeClassifier(max_depth=3, random_state=42)

In [9]:
#.score를 활용하여 DecisionTreeClassifier의 정확도 평가
print(f"DecisionTreeClassifier accuracy : {dt_model.score(X_train, y_train)}")

DecisionTreeClassifier accuracy : 0.9583333333333334


결정트리모델로 데이터를 학습 후에 평가지표를 도출하니 95.8%의 정확도가 나옴.

### RandomForestClassifier

In [10]:
#RandomForestClassifier를 활용하여 학습

rf_model = RandomForestClassifier(max_depth = 3, random_state = 42)
rf_model.fit(X_train, y_train)

RandomForestClassifier(max_depth=3, random_state=42)

In [11]:
#.score를 활용하여 RandomForestClassifier의 정확도 평가
print(f"RandomForestClassifier accuracy : {rf_model.score(X_train, y_train)}")

RandomForestClassifier accuracy : 0.9583333333333334


랜덤포레스트 모델로 데이터를 학습 후에 평가지표를 도출하니 95.8%의 정확도가 나옴.

### KNN

In [12]:
#KNeighborsClassifier를 활용하여 학습

knn_model = KNeighborsClassifier(n_neighbors = 3)
knn_model.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [13]:
#.score를 활용하여 KNeighborsClassifier의 정확도 평가
print(f"KNeighborsClassifier accuracy : {knn_model.score(X_train, y_train)}")

KNeighborsClassifier accuracy : 0.95


KNN 모델로 데이터를 학습 후에 평가지표를 도출하니 95.0%의 정확도가 나옴.

# 회귀 : 보스턴 집값

In [14]:
boston = pd.read_csv('보스턴 집값 데이터.csv')
boston.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  PTRATIO  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

        B  LSTAT  MEDV  CAT. MEDV  
0  396.90   4.98  24.0          0  
1  396.90   9.14  21.6          0  
2  392.83   4.03  34.7          1  
3  394.63   2.94  33.4          1  
4  396.90   5.33  36.2          1

In [15]:
boston.columns

Index(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX',
       'PTRATIO', 'B', 'LSTAT', 'MEDV', 'CAT. MEDV'],
      dtype='object')

In [16]:
#절대값으로 상관계수 비교
boston.corrwith(boston['MEDV']).abs()

CRIM         0.388305
ZN           0.360445
INDUS        0.483725
CHAS         0.175260
NOX          0.427321
RM           0.695360
AGE          0.376955
DIS          0.249929
RAD          0.381626
TAX          0.468536
PTRATIO      0.507787
B            0.333461
LSTAT        0.737663
MEDV         1.000000
CAT. MEDV    0.789789
dtype: float64

In [17]:
#boston의 독립변수와 종속변수 설정

boston_data = boston[['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX',
       'PTRATIO', 'B', 'LSTAT']]
boston_label = boston['MEDV']

In [18]:
#train_test_split을 활용하여 독립변수와 종속변수를 할당

X_train, X_test, y_train, y_test = train_test_split(boston_data, boston_label, test_size = 0.2, random_state = 42)

In [19]:
#LinearRegression를 활용하여 학습

lr_model = LinearRegression()
lr_model.fit(X_train, y_train)

LinearRegression()

In [20]:
#predict함수를 활용하여 예측
y_pred_lr = lr_model.predict(X_test)

In [21]:
#y_test와 예측값의 표준편차 구함, 작을수록 정확도 높아짐.
RMSE = metrics.mean_squared_error(y_test, y_pred_lr, squared = False)
print(RMSE)

4.928602182665361


In [22]:
#.score를 활용하여 LinearRegression의 정확도 평가
print(f"LinearRegression accuracy : {lr_model.score(X_train, y_train)}")

LinearRegression accuracy : 0.7508856358979673


선형회귀 모델로 데이터를 학습 후에 평가지표를 도출하니 75.1%의 정확도가 나옴.